Add DeprecationRestHandler to automatically log deprecated REST calls

pickypg opened the pull request:
This adds a new proxy for `RestHandler`s and `RestController`s so that requests made
to deprecated REST APIs can be automatically logged in the ES logs via the
`DeprecationLogger` as well as via a "Warning" header (RFC-7234) for all responses.

This PR does not touch any APIs in order to make this as easy to backport as possible.

Closes #17687

jasontedor commented on core/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -480,6 +480,34 @@ public static boolean validFileNameExcludingAstrix(String fileName) {
+     *
+     * @return {@code true} if the {@code value} is not obviously wrong.
+     */
+    public static boolean validHeaderValue(String value) {
```
I don't think this method belongs on `Strings`, especially since it's only used in only place. I'd be fine with it as a static package-private method on `DeprecationRestChannel` so it's visible for testing.

pickypg replied:
I honestly think that it _should_ be used in other places.

jasontedor replied:
That's probably correct, but right now it's not and I don't think `Strings` is the right place anyway.

pickypg replied:
Moved to `DeprecationRestChannel` and also added `String requireValidHeader(String value)` that behaves similar to `Objects.requireNonNull` to use for all creations. I did leave it as `public` as it's used by `DeprecationRestHandler` (even though they're in the same package).

jasontedor commented on core/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java:
```diff
@@ -404,6 +405,11 @@ public void apply(Settings value, Settings current, Settings previous) {
+                    IndexingMemoryController.MIN_INDEX_BUFFER_SIZE_SETTING,
+                    IndexingMemoryController.MAX_INDEX_BUFFER_SIZE_SETTING,
+                    IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING,
+                    IndexingMemoryController.SHARD_MEMORY_INTERVAL_TIME_SETTING
```
I'm confused by these changes? What's happening here?

pickypg replied:
It was a local rebase gone crazy. Those weren't from here.

jasontedor commented on core/src/main/java/org/elasticsearch/common/settings/Setting.java:
```diff
@@ -523,9 +523,9 @@ public static TimeValue parseTimeValue(String s, TimeValue minValue, String key)
+    public static Setting<ByteSizeValue> byteSizeSetting(String key, ByteSizeValue defaultValue, ByteSizeValue minValue, ByteSizeValue maxValue,
                                                          Property... properties) {
-        return byteSizeSetting(key, (s) -> value.toString(), minValue, maxValue, properties);
+        return byteSizeSetting(key, (s) -> defaultValue.toString(), minValue, maxValue, properties);
```
These too, they look like they are from #17778 which I reviewed earlier today?

jasontedor commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -1404,7 +1404,7 @@ private final EngineConfig newEngineConfig(EngineConfig.OpenMode openMode, Trans
             threadPool, indexSettings, warmer, store, deletionPolicy, indexSettings.getMergePolicy(),
             mapperService.indexAnalyzer(), similarityService.similarity(mapperService), codecService, shardEventListener, translogRecoveryPerformer, indexCache.query(), cachingPolicy, translogConfig,
-            indexSettings.getSettings().getAsTime(IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING, IndexingMemoryController.SHARD_DEFAULT_INACTIVE_TIME));
+            IndexingMemoryController.SHARD_INACTIVE_TIME_SETTING.get(indexSettings.getSettings()));
```
This too is confusing.

jasontedor commented on core/src/main/java/org/elasticsearch/indices/IndexingMemoryController.java:
```diff
@@ -50,22 +52,27 @@
 
     /** How much heap (% or bytes) we will share across all actively indexing shards on this node (default: 10%). */
-    public static final String INDEX_BUFFER_SIZE_SETTING = "indices.memory.index_buffer_size";
+    public static final Setting<ByteSizeValue> INDEX_BUFFER_SIZE_SETTING = Setting.byteSizeSetting("indices.memory.index_buffer_size", "10%", Property.NodeScope);
```
More that shouldn't be here.

jasontedor commented on core/src/test/java/org/elasticsearch/indices/IndexingMemoryControllerTests.java:
```diff
@@ -76,7 +74,7 @@
         public MockController(Settings settings) {
             super(Settings.builder()
-                            .put(SHARD_MEMORY_INTERVAL_TIME_SETTING, "200h") // disable it
+                            .put("indices.memory.interval", "200h") // disable it
```
More from #17778.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/DeprecationRestHandlerTests.java:
```diff
@@ -0,0 +1,88 @@
+ */
+public class DeprecationRestHandlerTests extends ESTestCase {
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
We prefer using `expectThrows` instead of this JUnit rule. Can you convert to using that pattern?

pickypg replied:
Of course. Noticed this one first :)

jasontedor replied:
Yeah, it's relatively new but it's the clear path forward especially with JUnit 5 coming with built-in support for the same.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestControllerTests.java:
```diff
@@ -34,9 +35,13 @@
+import static org.mockito.Mockito.verify;
 
 public class RestControllerTests extends ESTestCase {
-
```
I think we prefer the extra blank line between the definition of the class and the body of the class (and between the body of the class and the closing brace for the class). No, this isn't held to perfectly, but you'll see it across the codebase.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+    public static final String DEPRECATION_HEADER = "Warning";
+
+    /**
+     * The proxied canchannel.
```
I think these comments can be removed.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+     */
+    private final RestChannel channel;
+    /**
+     * The deprecation message.
```
I think these comments can be removed.

nik9000 replied:
If you make the variable's name `deprecationMessage` then, sure.

pickypg replied:
I prefer that name. I think I got the "Warning" header name stuck in my head early on. Renamed.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
If you do not provide *any* Javadoc, `inheritdoc` is [implied](http://docs.oracle.com/javase/6/docs/technotes/tools/solaris/javadoc.html#inheritingcomments).

pickypg replied:
I am aware of its implicit nature, but explicit documentation shows intent while implicit documentation can mean anything. I would strongly prefer to keep these here even though I know that it's implicit.

jasontedor replied:
> I am aware of its implicit nature, but explicit documentation shows intent while implicit documentation can mean anything. 

I'm not sure what you mean by this? The Javadoc for this method will have exactly the same documentation whether or not the `inheritdoc` tag is there.

> I would strongly prefer to keep these here even though I know that it's implicit.

It just adds source code clutter; I think that it should be removed. This is also inconsistent with how we do intend to do it everywhere else in the codebase (is it 100% perfect? probably not

nik9000 replied:
> It just adds source code clutter; I think that it should be removed.

+1

`@Override` is sufficient to say "I'm doing what the interface says." The biggest problem with this is that you have to jump to the overridden method to check its docs. `inheritdoc` doesn't solve that either.

pickypg replied:
> I'm not sure what you mean by this? The Javadoc for this method will have exactly the same documentation whether or not the inheritdoc tag is there.

It's not what appears in the theoretical javadoc jar, rather it's what it says in the code by having it there. Not having it there says one of two things:

1. I quickly wrote this code
2. I accept the same contract as the parent

> This is also inconsistent with how we do intend

I've never really taken it as intent versus just not happening (noting that the parent has zero javadoc except on one method). If we are actually intentionally not javadocing things, then I think that's both wrong and dangerous, but I will conform for inheritdoc.

jasontedor replied:
> It's not what appears in the theoretical javadoc jar, rather it's what it says in the code by having it there. Not having it there says one of two things:

> 1. I quickly wrote this code
> 2. I accept the same contract as the parent

If we're writing and reviewing code carefully, it only means the latter because those are the very straightforward rules of Javadocs. 

> If we are actually intentionally not javadocing things, then I think that's both wrong and dangerous

I completely disagree. Explicit is good when things are confusing, but not when things have straightforward rules.

For example, I note that you didn't use explicitly `this` in all cases when accessing a member variable. Why not? Probably because in those cases when it's not needed (i.e., there is no other variable that is shadowing the member variable), it is straightforward to understand.

> , but I will conform for inheritdoc.

There is no "conforming". It's better than we reach an understanding on this.

pickypg replied:
> If we're writing and reviewing code carefully
> It's better than we reach an understanding on this.

I disagree with the notion that the reviews make it acceptable to drop it because _only_ the submitter and reviewer(s) are aware of what happened, the reasoning, and any assumptions that goes into the changes. Particularly because of Guice, it's extremely annoying to jump through the code with undocumented-after-undocumented thing, having to dig into the weeds to find things that should be pretty simple. (This is more in regard to our general disregard for Javadoc, not `inheritdoc`)

I rarely document things for myself because I inherently know what I did (I'd hope, right? :)); it's intended for someone down the road that is not coming in with my biases and assumptions, or a total awareness of the entire codebase. In the case of `inheritdoc`, I agree it's _much_ looser than full documentation, but the explicit nature of it is far different as a reader to show that it's not supposed to change from parent behavior, especially in the context of the rest of our codebase. Furthermore, it suggests to any future reviewer to have someone changing the method to document their changes rather than just keep that as-is in a review somewhere. I got this way because I've inherited too many bad codebases (not suggesting that's what we have by any stretch) where they didn't document anything _and_ the code was bad/broken, which led to all sorts of pain whenever anything needed to be changed by a new member of the team. Applying `inheritdoc` also starts a good practice of actually adding Javadoc.

For me, it's very different from using/not using _syntax_ like `this` because that requires you to read the code either way. Documentation is supposed to help to give context about the contract without reading the raw implementation (e.g., side-effects, can it ever return `null`, when does it throw exceptions, etc.). It's like looking up JDK classes without looking up their implementation. For very advanced stuff, I certainly jump around the implementation, but for simple things it's nice to always know about explicit intent. I think our own inherent disagreement on when to use it shows that the implicit nature leads to different assumptions.

jasontedor replied:
> I disagree with the notion that the reviews make it acceptable to drop it because only the submitter and reviewer(s) are aware of what happened, the reasoning, and any assumptions that goes into the changes.

Your claim was that not having it there means one of two things: quickly written code, or we accept the same contract as the parent. I'm saying if we are writing and reviewing code carefully (which we are, right?), then it can not be the case that it's there because the code was written quickly. At least one of the writer and the reviewer (preferably both) have carefully considered the issue at hand.

> Particularly because of Guice, it's extremely annoying to jump through the code with undocumented-after-undocumented thing, having to dig into the weeds to find things that should be pretty simple.

This has nothing to do with whether or not `inheritdoc` is present.

> This is more in regard to our general disregard for Javadoc, not `inheritdoc`

I don't think we have "disregard" for it so much as we don't do a good job of using it when applicable.

I consider the rules around the inheriting of Javadocs here to be incredibly straightforward that they require no clarification. Do you *explicitly* cast every `int` parameter to a `long` when assigning an `int` to `long`? I don't, because the rules are straightforward.

> Applying inheritdoc also starts a good practice of actually adding Javadoc.

It's just clutter.

> Documentation is supposed to help to give context about the contract without reading the raw implementation (e.g., side-effects, can it ever return `null`, when does it throw exceptions, etc.).

We are not arguing over whether or not documentation should be present, we are arguing about cluttering the code with three lines that add nothing because they are otherwise implied.

pickypg replied:
I think we just disagree on the semantic value of it being explicit. Honestly, I still think we should have it, but I don't feel _that_ strongly about it since it is implied. Anything past this will be circular. I'm in the minority opinion and that's okay.

> Do you explicitly cast every int parameter to a long when assigning an int to long? 

Nope (except in a `ByteBuffer` situation, but I know you didn't mean those edge cases). But that's an implementation detail rather than a contract. I would even call out such an unnecessary cast in a code review, like you've done here. However, I think the semantics of casting `int` to `long` is different (it doesn't just add clutter, it creates questions) from explicitly specifying `inheritDoc` -- even though both are automatic expansions, doing so with a contract means something to me and at worst clutter. I've removed it since clearly it's not adding value for others.

jasontedor commented on core/src/test/java/org/elasticsearch/common/StringsTests.java:
```diff
@@ -60,4 +62,49 @@ public void testCleanTruncate() {
+         * Create a generator for characters [32, 126].
+         */
+        public ASCIIHeaderGenerator() {
+        super(asciiFromTo(32, 126));
```
The indentation is off here.

pickypg replied:
IntelliJ's folding hid this from me. Fixed (and moved).

jasontedor commented on core/src/test/java/org/elasticsearch/common/StringsTests.java:
```diff
@@ -60,4 +62,49 @@ public void testCleanTruncate() {
+         */
+        public ASCIIHeaderGenerator() {
+        super(asciiFromTo(32, 126));
+    }
```
The indentation is off here.

jasontedor commented on core/src/test/java/org/elasticsearch/rest/DeprecationRestChannelTests.java:
```diff
@@ -0,0 +1,122 @@
+ */
+public class DeprecationRestChannelTests extends ESTestCase {
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
Same thing, prefer `expectThrows`.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+    public DeprecationRestChannel(RestChannel channel, String warning) {
+        assert Strings.validHeaderValue(warning);
+
+        // required
```
Unneeded comment, it's clear from the code that they are required.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+    public void sendResponse(RestResponse response) {
+        response.addHeader(DEPRECATION_HEADER, warning);
+
+        // actually send the response
```
Unneeded comment.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -0,0 +1,108 @@
+     */
+    public DeprecationRestHandler(RestHandler handler, String warning, DeprecationLogger deprecationLogger) {
+        // it had better not be empty or you're deprecating it wrong
+        assert Strings.hasText(warning);
```
Is this right? Shouldn't it be `validHeaderValue`? And I don't think this should be an assertion, but a hard failure (assertions are disabled in production and if we are sending bad headers something is seriously wrong and we need to die hard).

pickypg replied:
I thought about it, but this is just logging. It's not until we get to the channel where it changes purpose and sends a header.

Though I do prefer that it fails fast instead of lazily later.

jasontedor replied:
> Though I do prefer that it fails fast instead of lazily later.

++

pickypg replied:
Though does this change your opinion of where that utility function lives?

jasontedor replied:
No, I still think that it should not be a method on the `Strings` class.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -0,0 +1,108 @@
+        // it had better not be empty or you're deprecating it wrong
+        assert Strings.hasText(warning);
+
+        // required
```
No comment necessary.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -0,0 +1,108 @@
+     */
+    @Override
+    public void handleRequest(RestRequest request, RestChannel channel) throws Exception {
+        // note the deprecation for ES logs
```
I think this method is clear enough without the comments.

jasontedor commented on core/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -480,6 +480,34 @@ public static boolean validFileNameExcludingAstrix(String fileName) {
+            char c = value.charAt(i);
+
+            // 32 = ' ' (31 = unit separator); 126 = '~' (127 = DEL)
+            // http://www.asciitable.com/ in case you need a reference
```
This comment can be removed, it's not authoritative, it might get broken, and it's easily discovered via search. What really matters is the RFC link which you've provided above already. :smile:

pickypg replied:
(Removed link, but method was moved)

jasontedor commented on core/src/main/java/org/elasticsearch/common/Strings.java:
```diff
@@ -480,6 +480,34 @@ public static boolean validFileNameExcludingAstrix(String fileName) {
+            return false;
+        }
+
+        for (int i = 0; i < value.length(); i++) {
```
This can probably just be simplified to `value.chars().allMatch(i -> i >= 32 && i <= 126);`

pickypg replied:
I think I'll do a second pass and look to make that kind of cleanup apply to all of the `Strings` functions as a master-only PR afterward?

jasontedor commented on core/src/test/java/org/elasticsearch/rest/RestControllerTests.java:
```diff
@@ -96,4 +101,63 @@ void executeHandler(RestRequest request, RestChannel channel) throws Exception {
+        public void handleRequest(RestRequest request, RestChannel channel) throws Exception {
+            throw new UnsupportedOperationException("not tested");
+        }
+    }
```
Blank line here.

jasontedor commented on core/src/test/java/org/elasticsearch/common/StringsTests.java:
```diff
@@ -19,6 +19,8 @@
+
 import org.elasticsearch.test.ESTestCase;
 
 public class StringsTests extends ESTestCase {
```
Wherever these end up, I think that there should be a test for the null string and the empty string being valid header values because those are edge cases that are easily messed up in some refactoring.

pickypg replied:
Agreed. Added.

jasontedor commented:
@pickypg Which endpoints is this going to target in 2.x?

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -112,7 +114,51 @@ public synchronized void registerFilter(RestFilter preProcessor) {
+     * @param handler The handler to actually execute the behavior
+     * @param warning The warning to log and send as a header in the response
+     */
+    public void registerDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler, String warning) {
```
Everything else is `Deprecation` so I think this should be too. Also, I think that `registerHandlerAsDeprecationHandler` might be a clearer name since we are really wrapping a handler.

jasontedor replied:
Do we need all three of the register methods? Can we get by with one in this PR until we see the need for the others?

pickypg replied:
I looked into it and it appears that `AbstractComponent` gives every component a `DeprecationLogger`, which I was unaware of when I wrote the three variants. So:

> Do we need all three of the register methods?

Nope. We only need the `DeprecationLogger` one (same for constructor below).

pickypg replied:
> Also, I think that registerHandlerAsDeprecationHandler might be a clearer name since we are really wrapping a handler.

I thought I was getting at that without making it too wordy. Happy to change it though as it does seem weird to have "Deprecated" with everything else being "Deprecation".

jasontedor replied:
> I thought I was getting at that without making it too wordy.

The key is that you're not registering a `DeprecationRestHandler` as the name `registerDeprecatedHandler` implies. Instead, you're registering a handler that gets wrapped as a `DeprecationRestHandler` before registration.

I guess `registerAsDeprecationHandler` would be slightly less wordy than `registerHandlerAsDeprecationHandler` but the point still remains.

pickypg replied:
I'm okay with `registerHandlerAsDeprecationHandler`; I think it let's someone using it know what's going to happen best. It's just a mouthful. :)

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -112,7 +114,51 @@ public synchronized void registerFilter(RestFilter preProcessor) {
+     * @param warning The warning to log and send as a header in the response
+     */
+    public void registerDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler, String warning) {
+        registerHandler(method, path, new DeprecationRestHandler(handler, warning, handler.getClass()));
```
Should we assert that `handler` is not an instance of `DeprecationHandler`?

pickypg replied:
It seems reasonable to protect us. Added (to the one-remaining overload).

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -112,7 +114,51 @@ public synchronized void registerFilter(RestFilter preProcessor) {
+     * @param warning The warning to log and send as a header in the response
+     * @param logger The wrapped logger to use for notifying for deprecation
+     */
+    public void registerDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler, String warning,
```
Same comments as the other override.

pickypg replied:
Method removed.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -112,7 +114,51 @@ public synchronized void registerFilter(RestFilter preProcessor) {
+     * @param warning The warning to log and send as a header in the response
+     * @param logger The existing deprecation logger to use
+     */
+    public void registerDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler, String warning,
```
Same comments: name, assert.

jasontedor commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestHandler.java:
```diff
@@ -0,0 +1,108 @@
+     * @param deprecationLogger The deprecation logger
+     * @throws NullPointerException if any parameter is {@code null}
+     */
+    public DeprecationRestHandler(RestHandler handler, String warning, DeprecationLogger deprecationLogger) {
```
Do we need all these constructors? Can we just get away with one?

pickypg replied:
Same comment as above: no, we do not need them now that I know that `AbstractComponent` gives everything a `DeprecationLogger`.

jasontedor commented:
I left a lot of comments, some of them apply to other places but I refrained from repeating the comment. I think the approach is good, but it's tricky to review without seeing the uses.

pickypg commented:
> Which endpoints is this going to target in 2.x?

To start:

- `/_optimize` becoming `/_forcemerge`
- `/{index}/_optimize` becoming `/{index}/_forcemerge`.

The second of which is likely used by a lot of cron jobs out in the wild.

nik9000 commented on core/src/main/java/org/elasticsearch/rest/DeprecationRestChannel.java:
```diff
@@ -0,0 +1,130 @@
+ * the {@link #DEPRECATION_HEADER deprecation header} and a warning message to denote that the existing request/response is
+ * deprecated.
+ */
+public class DeprecationRestChannel implements RestChannel {
```
Personally I'd prefer if you made a `DelegatingRestChannel` that does nothing but delegate to an internal `channel` member. Then `DeprecationRestChannel` can extend **that**, overriding the one method that it needs to override. That way it is **super** obvious which methods are just delegated and which are intercepted.

This isn't a requirement, but I think it makes these kinds of things easier to read.

pickypg replied:
I agree with that and I almost did it, but I'm not sure who else might use that code later? Since you had the same thought, I'm more than happy to do it.

nik9000 replied:
It doesn't matter if someone else will use it - the fact that you made two classes makes it easier to read, I think. I'm not sure everyone will agree with that assessment but it doesn't matter. The delegating instance is easy to review because you only have to hold one concept in your head, "this delegates".

pickypg replied:
Agreed. Added.

pickypg commented:
> I think the approach is good, but it's tricky to review without seeing the uses.

Yeah, it's the particular downside to making backporting easy. :( In general though, the way that I expect to see it (armed with the knowledge that endpoints already have a `DeprecationLogger`):

```java
        controller.registerHandler(POST, "/_forcemerge", this);
        controller.registerHandlerAsDeprecationHandler(POST, "/_optimize", this,
                                                       "Deprecated endpoint [/_optimize] used, replaced by [/_forcemerge]",
                                                       deprecationLogger);
        controller.registerHandler(POST, "/{index}/_forcemerge", this);
        controller.registerHandlerAsDeprecationHandler(POST, "/{index}/_optimize", this,
                                                       "Deprecated endpoint [/{index}/_optimize] used, replaced by [/{index}/_forcemerge]",
                                                       deprecationLogger);
```

Not every API will be renamed, so it's not as simple as just giving it the new name and letting it generate the description for us.

pickypg commented:
@jasontedor / @nik9000 I've modified from review.

clintongormley commented:
> > Which endpoints is this going to target in 2.x?
> To start:
> 
> /_optimize becoming /_forcemerge
> /{index}/_optimize becoming /{index}/_forcemerge.

Hmmm I was hoping that this PR would capture anything that being sent to the deprecation log, and include that info in the headers.  We shouldn't have a separate set of deprecation rules for headers.

pickypg commented:
@clintongormley

> Hmmm I was hoping that this PR would capture anything that being sent to the deprecation log, and include that info in the headers. We shouldn't have a separate set of deprecation rules for headers.

Interesting, that didn't occur to me with the issue (the issue was purpose-driven :)). The only way for us to do that would be to intercept/listen to calls to the deprecation logger, then somehow associate them with the current request/response and add them as a separate warning header (which is the appropriate way to do it according to the RFC as well, which is a good thing).

Worse, we may not even be the node that does the deprecation logging (e.g., a cluster state change getting handled by the elected master, with the REST call being handled by a data node). That makes it a bigger change, but I think it's doable with a little research on my part by adding it to the transport header instead of even worrying about REST for the most part (because REST auto-fills from the transport headers).

That said, deprecating of REST endpoints still needs to be handled and acknowledged with each request because once we start handling it, we effectively ignore its origin and at the start of handling we don't even have the transport headers created yet, thus making it the exception to the rule. So I think that would go in tandem with this PR.

clintongormley commented:
> That said, deprecating of REST endpoints still needs to be handled and acknowledged with each request 

I'd use the deprecation logger for deprecated REST endpoints, same as we do for everything else. Then, when (if?) we manage to return deprecation warnings in headers, then we'll get REST endpoints for free.

uboness commented:
what's the status of this PR?

pickypg commented:
@uboness It was stalled, but it's back on track.

pickypg commented:
I have it working to get _all_ deprecation log statements now, but I have a few issues to work out.

```bash
curl -XPUT -i localhost:9200/test -d '{
>   "settings": {
>     "number_of_shards": 1,
>     "number_of_replicas": 0
>   },
>   "mappings": {
>     "type": {
>       "properties": {
>         "field" : {
>           "type" : "string"
>         }
>       }
>     }
>   }
> }'
HTTP/1.1 200 OK
Warning: The [string] field is deprecated, please use [text] or [keyword] instead on [field]
Content-Type: application/json; charset=UTF-8
Content-Length: 21

{"acknowledged":true}
```

Here's an example with multiple deprecations (I locally deprecated `/{index}/_search` using the method adding by this PR to make an easy example):

```bash
curl -XPOST -i localhost:9200/test/_search -d '{
>   "query": {
>     "bool": {
>       "must": [
>         {
>           "indices": {
>             "indices": [
>               "test"
>             ],
>             "query": {
>               "match_all": {}
>             }
>           }
>         }
>       ]
>     }
>   }
> }'
HTTP/1.1 200 OK
Warning: /_search is not really deprecated
Warning: indices query is deprecated. Instead search on the '_index' field
Content-Type: application/json; charset=UTF-8
Content-Length: 236

{"took":17,"timed_out":false,"_shards":{"total":1,"successful":1,"failed":0},"hits":{"total":1,"max_score":1.0,"hits":[{"_index":"test","_type":"type","_id":"1","_score":1.0,"_source":{
  "field": "value",
  "message": "raw text"
}
}]}}
```

It's pretty cool and it works pretty well. It's just a matter of squashing some bugs. /cc @clintongormley

pickypg commented:
This is ready for review.

pickypg commented:
@jaymode Would you mind taking a look?

jaymode commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java:
```diff
@@ -419,6 +420,21 @@ public final Boolean readOptionalBoolean() throws IOException {
+    /**
+     * Read a map of strings to string lists.
+     */
+    public Map<String, List<String>> readMapOfLists() throws IOException {
```
The name makes the method sound generic and I think the method may have more uses if it is generic, like:

```java
    public <T, R> Map<T, List<R>> readMapOfLists(StreamInputReader<T> keyReader, StreamInputReader<R> listReader) throws IOException {
        int size = readVInt();
        if (size == 0) {
            return Collections.emptyMap();
        }
        Map<T, List<R>> map = new HashMap<>(size);
        for (int i = 0; i < size; ++i) {
            map.put(keyReader.read(this), readList(listReader));
        }
        return map;
    }
```

The method above is not tied to string keys and string lists. The caller here would use:

```java
in.readMapOfLists(StreamInput::readString, StreamInput::readString);
```

pickypg replied:
I started doing that, but then I realized that it will make it impossible to backport to 2.x without rewriting it back to this (which is still a goal).

I do prefer this approach, particularly for the reader side, and will follow up with a separate PR to do it on master only.

jaymode replied:
Sounds good. Do you mind opening up a issue so it doesn't get lost?

jaymode commented on core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -20,12 +20,44 @@
+     * This is set once by the {@code Node} constructor, but it uses {@link CopyOnWriteArraySet} to ensure that
+     * tests can run in parallel.
+     */
+    private static final CopyOnWriteArraySet<ThreadContext> THREAD_CONTEXT = new CopyOnWriteArraySet<>();
```
I know why you did this, but I think more details in the comments would help ensure the reasoning is clear. Things like tests spawn multiple nodes with separate injectors but the classes come from the same classloader so static variables get shared between nodes.

I wonder if we can workaround the underlying issue in the test framework by creating each node in its own classloader inside of ```InternalTestCluster#buildNode``` so that statics don't get shared

pickypg replied:
https://github.com/elastic/elasticsearch/issues/18540

Added a comment to describe why it's necessary though.

jaymode commented on core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -20,12 +20,44 @@
+    public static void setThreadContext(ThreadContext threadContext) {
+        assert threadContext != null;
+
+        THREAD_CONTEXT.add(threadContext);
```
In case this is left as a set, can we add a check to throw an exception if the set already contains the ThreadContext

pickypg replied:
Added a check. On the bright side, it didn't trigger any test failures.

pickypg commented on core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java:
```diff
@@ -401,6 +401,21 @@ public void writeMap(@Nullable Map<String, Object> map) throws IOException {
+    /**
+     * Writes a map of strings to string lists.
+     */
+    public void writeMapOfLists(Map<String, List<String>> map) throws IOException {
```
Same as reader.

jaymode commented on core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:
```diff
@@ -234,32 +270,38 @@ default void restore() {
 
     static final class ThreadContextStruct {
-        private final Map<String,String> headers;
+        private final Map<String, String> persistentHeaders;
```
can this be ```requestHeaders```? I find it a bit confusing as response headers should also be persistent

jaymode commented on core/src/main/java/org/elasticsearch/common/util/concurrent/ThreadContext.java:
```diff
@@ -304,20 +361,26 @@ private ThreadContextStruct copyHeaders(Iterable<Map.Entry<String, String>> head
+            out.writeMapOfLists(responseHeaders);
+        }
+
+        private void writeOnlyResponseHeadersTo(StreamOutput out) throws IOException {
```
Can we fold this into `writeTo` and add a boolean to the signature (maybe `includeRequestHeaders`)? It seems like it just requires adding a if/else block

jaymode commented on core/src/main/java/org/elasticsearch/node/Node.java:
```diff
@@ -236,6 +236,10 @@ protected Node(Environment tmpEnv, Version version, Collection<Class<? extends P
             threadPool.setClusterSettings(injector.getInstance(ClusterSettings.class));
+
+            // adds the context to the DeprecationLogger so that it does not need to be injected everywhere
+            DeprecationLogger.setThreadContext(threadPool.getThreadContext());
```
can we move this to be right after the ThreadPool is created?

jaymode commented on core/src/main/java/org/elasticsearch/rest/RestController.java:
```diff
@@ -113,7 +114,28 @@ public synchronized void registerFilter(RestFilter preProcessor) {
+     * @param deprecationMessage The message to log and send as a header in the response
+     * @param logger The existing deprecation logger to use
+     */
+    public void registerHandlerAsDeprecationHandler(RestRequest.Method method, String path, RestHandler handler,
```
this is just personal preference so feel free to ignore it, but I like the name `registerDeprecatedHandler`

uboness replied:
++ on `registerDeprecatedHandler`

jasontedor replied:
There is a type called `DeprecationRestHandler`. That's what handlers passed to this method get wrapped into. But you don't and can't pass one of those to this method yet that's what the name `registerDeprecatedHandler` suggests you can and maybe should do. The method does not register deprecated handlers, it registers plain handlers as deprecated handlers. I think `registerAsDeprecatedHandler` is fine though. :smile:

jaymode commented:
@pickypg how does this work with calls that hit multiple different nodes and the response is combined? Some examples of what I am talking about are extenders of `TransportBroadcastAction`, `TransportBroadcastByNodeAction`, and `TransportNodesAction` (there may be others). 

My understanding is the response headers from the "sub requests" will get dropped except for the last one, which finishes the action.

pickypg commented:
@jaymode That's exactly what happens currently. There's _currently_ no scenario that I'm aware of where it makes any difference, but I can imagine that changing as we expand its usage.

I have been thinking about adding a `readResponseHeaders` method that combines the existing `ThreadContext`'s response headers with the responding one. What do you think?

jaymode commented:
> I have been thinking about adding a readResponseHeaders method that combines the existing ThreadContext's response headers with the responding one. What do you think?

Where would you do that? When a response comes back and you are reading the headers in, the execution is happening in a transport thread, which is most likely going to be different than the one that made the request so I am not sure where/how you will get the existing context.

I think this would need to be kept track of in the `AsyncAction` classes that send requests to all nodes and handle responses. Something like accumulate/merge headers on each response and then set in the context when finished.

pickypg commented:
You're right. I am taking a look at the `TransportService` itself to avoid forcing every `AsyncAction` to do this (manually).

pickypg commented:
@jaymode Alright, so I've gotten back to this and I've looked over what it will take to get it to do the scatter / merge in the async action (effectively). It's doable, but it adds complexity to do something that we currently won't take advantage of at all.

So what I want to do is this:

1. Add a flag so that response headers are only accumulated in the `ThreadContext` if they're expected (when the request originates from a REST call).
2. Add more tests. I already have an integration test in place that tests / proves that it will not accumulate sub-request's responses.

Merge that, then backport it to the 2.x branch. Then implement the sub-request handling on master only. What do you think?

jaymode commented:
> Add a flag so that response headers are only accumulated in the ThreadContext if they're expected (when the request originates from a REST call).

Do we really need this?

> Merge that, then backport it to the 2.x branch. Then implement the sub-request handling on master only

That works for me

jaymode commented on core/src/main/java/org/elasticsearch/transport/local/LocalTransportChannel.java:
```diff
@@ -104,7 +109,7 @@ private void sendResponseData(byte[] data) {
         targetTransport.workers().execute(() -> {
             ThreadContext threadContext = targetTransport.threadPool.getThreadContext();
-            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
+            try (ThreadContext.StoredContext ignore = threadContext.newStoredContext()) {
```
I don't think this should have been changed

pickypg replied:
Correct! Reverted that change.

jaymode commented on core/src/test/java/org/elasticsearch/rest/DeprecationHttpIT.java:
```diff
@@ -0,0 +1,200 @@
+    /**
+     * Attempts to do a scatter/gather request that expects unique responses per sub-request.
+     */
+    @AwaitsFix(bugUrl = "")
```
can we add text here/open up a issue?

pickypg replied:
Created https://github.com/elastic/elasticsearch/issues/19222 (and placed it appropriately)

jaymode commented on core/src/test/java/org/elasticsearch/rest/DeprecationHttpIT.java:
```diff
@@ -0,0 +1,200 @@
+        final String quotedCommaSeparatedIndices = Stream.of(indices).map((i) -> "\"" + i + "\"").collect(Collectors.joining(","));
+        final String body =
+            "{\"query\":{\"bool\":{\"filter\":[" +
+//                "{\"indices\":{\"indices\":[" + quotedCommaSeparatedIndices + "],\"query\":{\"" + TestDeprecatedQueryBuilder.NAME +  "\":{}}}}," +
```
is this needed?

pickypg replied:
No, I was using that to mess with my deprecation warnings. Removed from the committed code.

pickypg commented:
@jaymode Please take a second pass now that I have rebased it back onto master. I think it's ready to merge with all tests passing.

jaymode commented on core/src/main/java/org/elasticsearch/common/logging/DeprecationLogger.java:
```diff
@@ -47,8 +90,42 @@ public DeprecationLogger(ESLogger parentLogger) {
     }
+
+    /**
+     * Note: {@code threadContexts} <em>must</em> be a modifiable {@link Set} so that stale contexts can be removed.
```
I think it would be simpler to have the node remove its threadcontext from here on close

jaymode commented:
left one comment, other than that LGTM

frioux commented:
In case anyone in the internet superfuture runs into this like we did, this can pretty easily cause a long enough set of headers that nginx cannot even parse the response, and thus can cause pretty serious outages.  Would be great if it could be disabled entirely, but you can set `proxy_buffer_size` to something higher to resolve it at least temporarily.

jasontedor commented:
@frioux Can you share the contents of these large responses? I would be interested in seeing if there are headers that we need to de-duplicate (we do this already to some extent, this is part of what pushes my interest here).

frioux commented:
```HTTP/1.1 200 OK
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
Warning: 299 Elasticsearch-5.4.0-780f8c4 "The [string] field is deprecated, please use [text] or [keyword] instead on [REDACTED]" "Wed, 17 Jan 2018 19:03:29 GMT"
content-type: application/json; charset=UTF-8
content-encoding: gzip
transfer-encoding: chunked

1
```

Of course instead of `REDACTED` it was a distinct field for each thing.  12k is a lot of data to jam into headers ;)

jasontedor commented:
Okay; it is because it's a distinct field. Thanks for that, I will see what, if anything, can be done (I am doubtful because of the way these are collected).

frioux commented:
It would be good to limit this, because this is pretty gnarly to debug (I could only do it with a network trace) and anyone who puts `nginx` in front of ES would run into it.

jasontedor commented:
I agree with you. I opened #28301 and this work is assigned.
